### PR TITLE
Replace `daemonize` with a custom version

### PR DESCRIPTION
**Pull request by SohamG**
Replace [daemonize](https://github.com/thesharp/daemonize) with a custom version, working towards https://github.com/matrix-org/synapse/issues/1539

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

**Comment by SohamG** [comment]
@richvdh Here's the smaller PR with only `_daemonize.py` (from #6169)

File: .gitignore [hunk 1]
```diff
@@ -15,6 +15,7 @@ _trial_temp*/
 # stuff that is likely to exist when you run a server locally
 /*.db
 /*.log
+/*.out
```
Comment (richvdh) [inline .gitignore]: why is this change required?

File: synapse/app/_base.py [hunk 2]
```diff
@@ -21,7 +21,7 @@
 import sys
 import traceback
 
-from daemonize import Daemonize
```
Comment (richvdh) [inline synapse/app/_base.py]: we can remove `daemonize` from `python_dependencies.py`

File: synapse/app/_daemonize.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import sys
+import signal
+import atexit
+import traceback
+
+
+class Daemonize(object):
+
+    """
+    Daemonize object.
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: I see you've copied this from the original Daemonize implementation, but I don't think it's very useful. *obviously* it's a Daemonize object, but what does it do? if you can't think of anything useful to say, don't say anything at all.

File: synapse/app/_daemonize.py [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    Object constructor expects three arguments.
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: this is redundant too.

File: synapse/app/_daemonize.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    :param app: contains the application name which will be sent to syslog.
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: please could you make these follow the format of the docstrings elsewhere in the project, and include info on the types?

File: synapse/app/_daemonize.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    :param pid: path to the pidfile.
+    :param action: your custom function which will be executed after daemonization.
+    :param logger: use this logger object instead of creating new one, if provided.
+    :param chdir: change working directory if provided or /
+    :param outfile: File to redirect stdio to
+    :param void_stdio: Should stdio be sent to /dev/null or to outfile
+    """
+    def __init__(self, app, pid, action,
+                 verbose=False, logger=None,
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: `verbose` is unimplemented and unused: please get rid of it.

File: synapse/app/_daemonize.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    Object constructor expects three arguments.
+
+    :param app: contains the application name which will be sent to syslog.
+    :param pid: path to the pidfile.
+    :param action: your custom function which will be executed after daemonization.
+    :param logger: use this logger object instead of creating new one, if provided.
+    :param chdir: change working directory if provided or /
+    :param outfile: File to redirect stdio to
+    :param void_stdio: Should stdio be sent to /dev/null or to outfile
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: this is unclear: if I set it to to True, will it use the `outfile` or `/dev/null`?

File: synapse/app/_daemonize.py [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+    """
+    Daemonize object.
+
+    Object constructor expects three arguments.
+
+    :param app: contains the application name which will be sent to syslog.
+    :param pid: path to the pidfile.
+    :param action: your custom function which will be executed after daemonization.
+    :param logger: use this logger object instead of creating new one, if provided.
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: I suggest getting rid of `logger` and creating a new logger in the file, as all the other files in the project do.

File: synapse/app/_daemonize.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    :param chdir: change working directory if provided or /
+    :param outfile: File to redirect stdio to
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: I'd prefer it if we could avoid changing the functionality in the same PR as the refactor, particularly since we're hardcoding `homeserver.log` here.

For now I suggest replicating the existing behaviour (ie, removing the `outfile` and `void_stdio` params, and continuing to redirect to /dev/null). We can introduce an alternative output file in a later PR.

File: synapse/app/_daemonize.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        # Stop listening for signals that the parent process receives.
+        # This is done by getting a new process id.
+        # setpgrp() is an alternative to setsid().
+        # setsid puts the process in a new parent group and detaches its controlling terminal.
+        process_id = os.setsid()
+        if process_id == -1:
+            # Uh oh, there was a problem.
+            sys.exit(1)
+
+        if self.void_stdio:
```
Comment (richvdh) [inline synapse/app/_daemonize.py]: I think you've got this condition backwards fwiw

**Comment by richvdh** [comment]
going to close this until there's some progress on it.
